### PR TITLE
Ledger export upgrades

### PR DESCRIPTION
**Pull request by Sventimir**
Related PR group:
* #14889
* #14917
* #14934

This PR provides more upgrades for the ledger export feature.

Explain your changes:
* Insert explicit yields between expensive operations to minimise the risk of node spendignto much time in GraphQL computation.
* Include the protocol state in the exported config.
* Allow users to choose any block from the transition frontier to base the export on.
* Update the relevant benchmarks.

Explain how you tested your changes:
* Performed the ledger export from mainnet.
* Ran the benchmarks:

```
┌─────────────────────────────┬──────────┬──────────┬──────────┬──────────┬────────────┐
│ Name                        │ Time/Run │  mWd/Run │ mjWd/Run │ Prom/Run │ Percentage │
├─────────────────────────────┼──────────┼──────────┼──────────┼──────────┼────────────┤
│ parse_runtime_config        │    3.30s │ 614.90Mw │ 171.17Mw │ 171.17Mw │     61.27% │
│ serialize_runtime_config    │    1.69s │ 136.76Mw │ 251.67Mw │  80.26Mw │     31.34% │
│ convert_accounts_for_config │    5.39s │ 476.33Mw │  14.88Mw │  14.88Mw │    100.00% │
└─────────────────────────────┴──────────┴──────────┴──────────┴──────────┴────────────┘
```

Checklist:

- [x] Dependency versions are unchanged
  - Notify Velocity team if dependencies must change in CI
- [x] Modified the current draft of release notes with details on what is completed or incomplete within this project
- [x] Document code purpose, how to use it
  - Mention expected invariants, implicit constraints
- [x] Tests were added for the new behavior
  - Document test purpose, significance of failures
  - Test names should reflect their purpose
- [x] All tests pass (CI will check this if you didn't)
- [x] Serialized types are in stable-versioned modules

### CODE REVIEW

**Comment by Sventimir** [comment]
!ci-build-me

File: src/lib/runtime_config/runtime_config.ml [hunk 1]
```diff
@@ -1253,8 +1252,21 @@ let gen =
...
+  let open Async.Deferred.Result.Let_syntax in
+  let yield = Async_unix.Scheduler.yield_every ~n:100 |> Staged.unstage in
+  let%bind accounts =
+    Mina_base.Ledger.Any_ledger.M.to_list ledger
+    |> Async.Deferred.map ~f:Result.return
+  in
+  let%map accounts =
+    deferred_list_fold ~init:[]
+      ~f:(fun acc el ->
+        let%bind () = yield () |> Async.Deferred.map ~f:Result.return in
```
Comment (georgeee) [inline src/lib/runtime_config/runtime_config.ml:1264]:
```suggestion
        let%bind () = yield () >>| Result.return
```


It'd require replace `let%bind` above with `let%bind.Deferred.Result` and removal of `open Deferred.Result.Let_syntax`

I believe it's cleaner, though it's a NIT.

**Review by georgeee — APPROVED** [review]

File: src/lib/mina_graphql/mina_graphql.ml [hunk 2]
```diff
@@ -4361,12 +4361,15 @@ module Queries = struct
...
                     next_epoch.ledger.hash ) ) ;
-            let%map new_config =
+            let%bind new_config =
               Runtime_config.make_fork_config ~staged_ledger ~global_slot
                 ~staking_ledger ~staking_epoch_seed ~next_epoch_ledger
                 ~next_epoch_seed ~blockchain_length ~protocol_state
                 runtime_config
             in
+            let%map () =
+              Async_unix.Scheduler.yield () |> Deferred.map ~f:Result.return
```
[the remark this answers is not in the document]
Comment (Sventimir) [reply src/lib/mina_graphql/mina_graphql.ml]: How about replacing just this one line with the following?
```
          let open Async.Deferred.Infix in
          Async_unix.Scheduler.yield () >>| Result.return
```

File: src/lib/runtime_config/runtime_config.ml (hunk 1 above)
Comment (Sventimir) [reply src/lib/runtime_config/runtime_config.ml:1264]: Proposed a fix in [29131a6](https://github.com/MinaProtocol/mina/pull/14889/commits/29131a603048b1f96e55205eeab45cd738d97bdd).

File: src/lib/mina_graphql/mina_graphql.ml [hunk 3]
```diff
@@ -4292,85 +4261,110 @@ module Queries = struct
...
+              Mina_lib.best_chain_block_by_state_hash mina state_hash
+              |> Deferred.return
+          | None, Some block_height ->
+              Mina_lib.best_chain_block_by_height mina
+                (Unsigned.UInt32.of_int block_height)
+              |> Deferred.return
+          | Some _, Some _ ->
+              Deferred.Result.fail "Cannot specify both state hash and height"
+        in
+        let block = Transition_frontier.Breadcrumb.(block breadcrumb) in
```
Comment (mrmr1993) [inline src/lib/mina_graphql/mina_graphql.ml:4296]: Prefer qualified name without brackets.

File: src/lib/mina_graphql/mina_graphql.ml [hunk 4]
```diff
@@ -4292,85 +4261,110 @@ module Queries = struct
...
+              | `Active breadcrumb ->
+                  return breadcrumb )
+          | Some state_hash_base58, None ->
+              let%bind state_hash =
+                State_hash.of_base58_check state_hash_base58
+                |> Result.map_error ~f:Error.to_string_hum
+                |> Deferred.return
+              in
+              Mina_lib.best_chain_block_by_state_hash mina state_hash
+              |> Deferred.return
```
Comment (mrmr1993) [inline src/lib/mina_graphql/mina_graphql.ml:4288]: This might be cleaner as a pipeline, using `>>=` in place of the bind.

**Comment by Sventimir** [comment]
!ci-build-me

File: src/lib/runtime_config/runtime_config.ml (hunk 1 above)
Comment (mrmr1993) [reply src/lib/runtime_config/runtime_config.ml:1264]: That fix wasn't applied here

**Review by dkijania — APPROVED** [review]

File: src/lib/runtime_config/runtime_config.ml (hunk 1 above)
Comment (Sventimir) [reply src/lib/runtime_config/runtime_config.ml]: Sorry, applied in 6f44a34e98.

File: src/lib/mina_graphql/mina_graphql.ml (hunk 3 above)
Comment (Sventimir) [reply src/lib/mina_graphql/mina_graphql.ml]: Done in 6f44a34e98.

File: src/lib/mina_graphql/mina_graphql.ml (hunk 4 above)
Comment (Sventimir) [reply src/lib/mina_graphql/mina_graphql.ml]: Not sure if that's really cleaner, but here you are: 6f44a34e98.

**Comment by Sventimir** [comment]
!ci-build-me

**Review by georgeee — APPROVED** [review]

**Comment by stevenplatt** [comment]
!ci-build-me

**Comment by Sventimir** [comment]
!ci-build-me

File: src/lib/runtime_config/runtime_config.ml [hunk 5]
```diff
@@ -1283,16 +1301,34 @@ let make_fork_config ~staged_ledger ~global_slot ~blockchain_length
...
+        Some (Unsigned.UInt32.to_int protocol_constants.slots_per_sub_window)
+    ; genesis_state_timestamp =
+        Some (Block_time.to_string protocol_constants.genesis_state_timestamp)
+    }
+  in
   let fork =
     Fork_config.
       { previous_state_hash =
-          Mina_base.State_hash.to_base58_check protocol_state_hash
+          Mina_base.State_hash.to_base58_check
```
Comment (ghost-not-in-the-shell) [inline src/lib/runtime_config/runtime_config.ml:1322]: This should actually be the `state_hash` instead of `previous_state_hash`

**Pull request merged**